### PR TITLE
Serving all competitions leagues and matches

### PR DESCRIPTION
**Pull request by Janostefano**

### CODE REVIEW

File: src/test/groovy/com/notifball/notifball/TeamsControllerTest.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+@AutoConfigureMockMvc
+class TeamsControllerTest extends BaseE2ESpec {
+
+    def setup() {
+        stubTeamsDataProvider()
+    }
+
+    @Autowired
+    TeamsController teamsController
```
Comment (iwko) [inline src/test/groovy/com/notifball/notifball/TeamsControllerTest.groovy]: you do not need this

File: src/test/groovy/com/notifball/notifball/CompetitionsControllerTest.groovy [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    def setup() {
+        stubCompetitionsDataProvider()
+    }
+
+    @Autowired
+    protected MockMvc mvc
+
+    @Autowired
+    CompetitionsController competitionsController
```
Comment (iwko) [inline src/test/groovy/com/notifball/notifball/CompetitionsControllerTest.groovy]: You do not need thisl

File: src/test/groovy/com/notifball/notifball/CompetitionsControllerTest.groovy [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+    def "should serve all competitions"() {
+        given:
+        HttpEntity entity = new HttpEntity<>()
+        def response = restTemplate.exchange("${footballDataHost}/v2/competitions", GET, entity, CompetitionDto.CompetitionsDto.class)
```
Comment (iwko) [inline src/test/groovy/com/notifball/notifball/CompetitionsControllerTest.groovy]: Why do you want to take data from external API? In this scenario you should put data in DB using CompetitionRepository.save

File: src/test/groovy/com/notifball/notifball/MatchesControllerTest.groovy [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    @Autowired
+    protected MockMvc mockMvc
+
+    @Autowired
+    MatchesController matchesController
+
+    def "should serve all matches from db"() {
+        given:
+        HttpEntity entity = new HttpEntity<>()
+        def response = restTemplate.exchange("${footballDataHost}/v2/competitions/2021/matches?season=2019", HttpMethod.GET, entity, MatchDto.MatchesDto.class)
```
Comment (iwko) [inline src/test/groovy/com/notifball/notifball/MatchesControllerTest.groovy]: As the last argument you should use only simple `List`

File: src/main/java/com/notifball/notifball/api/CompetitionDto.java [hunk 5]
```diff
@@ -2,77 +2,94 @@
 
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.notifball.notifball.domain.competition.Competition;
 
 import java.util.List;
 
+
+@JsonSerialize
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionDto.java]: done

File: src/main/java/com/notifball/notifball/api/CompetitionDto.java [hunk 6]
```diff
@@ -2,77 +2,94 @@
...
         this.area = area.getArea();
-        this.startDate = season.getStartDate();
-        this.endDate = season.getEndDate();
-        this.currentMatchday = season.getCurrentMatchday();
+        this.startDate = currentSeason.getStartDate();
+        this.endDate = currentSeason.getEndDate();
+        this.currentMatchday = currentSeason.getCurrentMatchday();
+    }
+
+    public CompetitionDto() {
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionDto.java]: yes indeed

File: src/main/java/com/notifball/notifball/api/CompetitionsController.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import com.notifball.notifball.infrastructure.competition.PersistentCompetition;
+import com.notifball.notifball.infrastructure.match.PersistentMatch;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionsController.java]: done

File: src/main/java/com/notifball/notifball/api/CompetitionsController.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    private CompetitionRepository competitionRepository;
+
+    public CompetitionsController(CompetitionRepository repository) {
+
+        this.competitionRepository = repository;
+    }
+
+    @GetMapping(value = "", produces = MediaType.APPLICATION_JSON_VALUE)
+    public @ResponseBody
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionsController.java]: done

File: src/main/java/com/notifball/notifball/api/MatchDto.java [hunk 9]
```diff
@@ -6,179 +6,237 @@
...
-        private String utcDate;
-        private String status;
-        private String matchday;
-        private String stage;
-        private String group;
-        private String lastUpdated;
-        private ScoreDto score;
-        private TeamsDto.TeamDto homeTeam;
-        private TeamsDto.TeamDto awayTeam;
+    public MatchDto(String externalId, String utcDate, String status,
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/MatchDto.java]: done

File: src/main/java/com/notifball/notifball/api/CompetitionsController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionsController.java]: do we need to give type in this annotation?

File: src/main/java/com/notifball/notifball/api/CompetitionDto.java [hunk 10]
```diff
@@ -2,77 +2,94 @@
...
+        super();
     }
 
+    public CompetitionDto(String externalId,
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionDto.java]: one is to create an object from external api's data, the second one is for our api which takes different data from our db

File: src/main/java/com/notifball/notifball/api/MatchDto.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/MatchDto.java]: one is to create an object from external api's data, the second one is for our api which takes different data from our db

File: src/main/java/com/notifball/notifball/api/CompetitionsController.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    List<CompetitionDto> getAllCompetitions() {
+        return competitionRepository.findAll().
```
[the remark this answers is not in the document]
Comment (Janostefano) [reply src/main/java/com/notifball/notifball/api/CompetitionsController.java]: where should i create this class?

File: src/main/java/com/notifball/notifball/api/CompetitionDto.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (iwko) [reply src/main/java/com/notifball/notifball/api/CompetitionDto.java]: Removed

File: src/main/java/com/notifball/notifball/api/MatchDto.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (iwko) [reply src/main/java/com/notifball/notifball/api/MatchDto.java]: Resolved

**Pull request merged**